src/datasets/arrow_dataset.py

```diff
@@ -815,10 +806,12 @@ def from_dict(
         if features is not None:
             mapping = features.encode_batch(mapping)
         mapping = {
-            col: OptimizedTypedSequence(data, type=features.type[col].type if features is not None else None, col=col)
+            col: OptimizedTypedSequence(data, type=features[col] if features is not None else None, col=col)
```

Comment:
TypedSequence and OptimizedTypedSequence now take HF feature types as input instead of pyarrow types.

This is because it should be able to cast string to struct for Image and Audio for example, and such casting is defined by the HF feature types